hashintel/hash
# H-255: Implement root pagination for structural queries

## 🌟 What is the purpose of this PR?

To query in a more performant way, this adds pagination to the subgraph roots.

## 🔍 What does this change?

Allows two query parameters:
- `limit`: The number of roots to be returned from the graph
- `after`: Omit every entry returned before this cursor
Returns a new header:
- `Link` contains a link to the next page. Will only be returned if `limit` was passed. Please note [the format of the `Link` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Link).

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues

Due to the way how the `"latest"` filter for ontology type is implemented, pagination is not supported if ontology types are filtered by their latest version. See H-1442.

## ❓ How to test this?

- Query a few types or entities with `?limit=X`
- Make sure the `Link` header is returned and at most `X` items
- Use the provided link and query the next page
